custom_components/tuya_local/devices/cecotec_readywarm_8400.yaml

```diff
@@ -17,13 +17,12 @@ primary_entity:
           conditions:
             - dps_val: "null"
```

Comment:
It should have been the "null" mapping that was removed, not the "off" mapping, as the "null" value is not one of the value values for this dp.

Reply:
Ok so if I remove the null mapping from the constraint - then the Fan Only option disappears.
if I make the Fan Only dps_val == "off" then we are back to the original behaviour.

I merged your latest commit with the changes to hidden values from devices_config.py.  
I have also reverted to your commit for this heater.
It doesnt seem to have fixed the issue.

Reply:
With latest changes to devices_confiug.py the logs looks like:
`024-12-30 14:41:39.393 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 4 value from off to none
2024-12-30 14:41:39.393 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Considering mapping for range of preset_mode
2024-12-30 14:41:39.400 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 1 to True from fan_only
2024-12-30 14:41:39.400 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Considering mapping for range of hvac_mode
2024-12-30 14:41:39.401 DEBUG (MainThread) [custom_components.tuya_local.device] Fan Heater new pending updates: {"4": {"value": "none", "updated_at": 1735569699.4007902, "sent": false}, "1": {"value": true, "updated_at": 1735569699.4007902, "sent": false}}
2024-12-30 14:41:39.407 DEBUG (MainThread) [custom_components.tuya_local.device] Fan Heater sending dps update: {"1": true, "4": "none"}
2024-12-30 14:41:39.538 DEBUG (MainThread) [custom_components.tuya_local.device] Fan Heater received {"1": false, "2": 25, "3": 22, "4": "off", "8": false, "19": "0h", "20": 0, "full_poll": true}
2024-12-30 14:41:39.543 WARNING (MainThread) [custom_components.tuya_local.climate] cecotec_readywarm_8400.yaml/climate: Unrecognised HVAC Mode of True ignored
2024-12-30 14:41:39.545 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 8 value from False to off
2024-12-30 14:41:39.546 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 8 value from False to off
2024-12-30 14:41:39.549 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 19 value from 0h to cancel
2024-12-30 14:41:39.549 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 19 value from 0h to cancel
2024-12-30 14:41:39.652 DEBUG (MainThread) [custom_components.tuya_local.device] Fan Heater received {"1": true, "full_poll": false}
2024-12-30 14:41:39.656 WARNING (MainThread) [custom_components.tuya_local.climate] cecotec_readywarm_8400.yaml/climate: Unrecognised HVAC Mode of True ignored
2024-12-30 14:41:39.657 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 8 value from False to off
2024-12-30 14:41:39.657 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 8 value from False to off
2024-12-30 14:41:39.658 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 19 value from 0h to cancel
2024-12-30 14:41:39.658 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 19 value from 0h to cancel`

Reply:
To be honest I am not insisting anything as I am well out of my debt at this stage.  

The entire null idea came from another heater implementation that seemed to work without error - `vtaplus_axial_fanheater.yaml` - , but I agree appears to be a work around something to do with hidden values - as it makes little sense.

The last change was a misunderstanding on my part of what you told me. And I will revert it now.

The logs I have posted at this stage are of the functionality as it stands with your changes.

The first set of logs I posted  were with your iteration of the yaml files before your changes to devices_config,.py and the second set of logs are post the changes to devices_config.py.

Reply:
I reverted the change as per previous comments but also seem to have found a solution - by assigning the fan only state to be the default value of of a "true" DP1.  Constraints seems to be applied after - and it all appears to work happily.

@make-all  can you advise if that makes any sense to you?

logs:

`2024-12-31 11:25:53.507 DEBUG (MainThread) [custom_components.tuya_local.device] Fan Heater received {"1": false, "2": 25, "3": 22, "4": "off", "8": false, "19": "0h", "20": 0, "full_poll": true}
2024-12-31 11:25:53.510 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 19 value from 0h to cancel
2024-12-31 11:25:53.511 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 19 value from 0h to cancel
2024-12-31 11:25:53.512 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 1 value from False to off
2024-12-31 11:25:53.513 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 1 value from False to off
2024-12-31 11:25:53.514 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 4 value from off to none
2024-12-31 11:25:53.514 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 4 value from off to none
2024-12-31 11:25:53.514 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 8 value from False to off
2024-12-31 11:25:53.514 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 8 value from False to off
2024-12-31 11:26:02.511 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 1 to True from fan_only
2024-12-31 11:26:02.512 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Considering mapping for range of hvac_mode
2024-12-31 11:26:02.512 DEBUG (MainThread) [custom_components.tuya_local.device] Fan Heater new pending updates: {"1": {"value": true, "updated_at": 1735644362.512235, "sent": false}}
2024-12-31 11:26:02.514 DEBUG (MainThread) [custom_components.tuya_local.device] Fan Heater sending dps update: {"1": true}
2024-12-31 11:26:02.640 DEBUG (MainThread) [custom_components.tuya_local.device] Fan Heater received {"1": true, "full_poll": true}
2024-12-31 11:26:02.642 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 19 value from 0h to cancel
2024-12-31 11:26:02.642 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 19 value from 0h to cancel
2024-12-31 11:26:02.644 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 1 value from True to fan_only
2024-12-31 11:26:02.644 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 1 value from True to fan_only
2024-12-31 11:26:02.645 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 4 value from off to none
2024-12-31 11:26:02.645 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 4 value from off to none
2024-12-31 11:26:02.646 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 8 value from False to off
2024-12-31 11:26:02.647 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 8 value from False to off
2024-12-31 11:26:02.756 DEBUG (MainThread) [custom_components.tuya_local.device] Fan Heater received {"1": true, "2": 25, "3": 22, "4": "off", "8": false, "19": "0h", "20": 0, "full_poll": false}
2024-12-31 11:26:02.758 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 19 value from 0h to cancel
2024-12-31 11:26:02.758 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 19 value from 0h to cancel
2024-12-31 11:26:02.759 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 1 value from True to fan_only
2024-12-31 11:26:02.759 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 1 value from True to fan_only
2024-12-31 11:26:02.760 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 4 value from off to none
2024-12-31 11:26:02.760 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 4 value from off to none
2024-12-31 11:26:02.760 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 8 value from False to off
2024-12-31 11:26:02.761 DEBUG (MainThread) [custom_components.tuya_local.helpers.device_config] Fan Heater: Mapped dps 8 value from False to off`

Reply:
I think that would work with the latest version of the code. The old code would have skipped processing the conditions if it found a value outside of the conditions, but I realised that some configs are already trying to set default values like that since there is no option for a default condition, so it is better to continue to process the conditions.